metrics: add BLEU

ydcjeff opened the pull request:
## What does this PR do?

<!--
Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.
-->

Fixes #1301 (issue)

# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests? 
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

<!-- For CHANGELOG separate each item in unreleased section by a blank line to reduce collisions -->

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

pep8speaks commented:
Hello @ydcjeff! Thanks for opening this PR. 

* In the file [`pytorch_lightning/metrics/bleu.py`](https://github.com/PyTorchLightning/pytorch-lightning/blob/589d9b79b62e8577d67e480a99e93a3008637a17/pytorch_lightning/metrics/bleu.py):

> [Line 43:52](https://github.com/PyTorchLightning/pytorch-lightning/blob/589d9b79b62e8577d67e480a99e93a3008637a17/pytorch_lightning/metrics/bleu.py#L43): [E203](https://duckduckgo.com/?q=pep8%20E203) whitespace before ':'
> [Line 126:120](https://github.com/PyTorchLightning/pytorch-lightning/blob/589d9b79b62e8577d67e480a99e93a3008637a17/pytorch_lightning/metrics/bleu.py#L126): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (130 > 119 characters)



Do see the [Hitchhiker's guide to code style](https://goo.gl/hqbW4r)

ydcjeff commented:
Hello @justusschock,
I have opened the PR. Also, where should I put my code under metrics package as I wrote it in a separate file?

SkafteNicki commented:
Since bleu is a metric specific to nlp, could you move your code to a file called `nlp.py`. I guess we down the line will get more field-specific metrics.
Additionally,
* add link to metric in this file https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/metrics/__init__.py
* add reference in the docs, in this file https://github.com/PyTorchLightning/pytorch-lightning/blob/master/docs/source/metrics.rst

justusschock commented on pytorch_lightning/metrics/bleu.py:
```diff
@@ -0,0 +1,143 @@
+    return bleu
+
+
+# t = "the FAST brown fox jumped over the lazy dog"
```
can you remove these lines?

justusschock commented on pytorch_lightning/metrics/bleu.py:
```diff
@@ -0,0 +1,143 @@
+import re
+
+
+def math_log(num: int):
```
can we torchify this by using torch.log?

ydcjeff replied:
Okay, I will convert to Torch.

justusschock commented on pytorch_lightning/metrics/bleu.py:
```diff
@@ -0,0 +1,143 @@
+        ref_len: the best match length of text of the reference lists with translated text
+        translate_len: the length of translated text
+    """
+    translate_len = len(translate_list)
```
this also needs to be a pure pytorch implementation.

ydcjeff replied:
so the inputs to the `bleu_score` should all be torch.Tensor?
Not string?

justusschock commented on pytorch_lightning/metrics/bleu.py:
```diff
@@ -0,0 +1,143 @@
+    """
+    ngram_counter = Counter()
+
+    for i in range(1, num_grams + 1):
```
is there a more efficient way to do this than two nested for loops?

ydcjeff replied:
Sorry, I can't think of other than nested for loops. Can you help for a more efficient way?

justusschock commented on pytorch_lightning/metrics/bleu.py:
```diff
@@ -0,0 +1,143 @@
+    for i in range(n):
+        precision_scores[i] = numerator[i] / denominator[i]
+
+    geometric_mean = math.exp(sum((1.0 / n) * math_log(p) for p in precision_scores))
```
convert this to torch as well.

ydcjeff replied:
Okay, I will do it.

ydcjeff commented:
I have refactored with torch.Tensor, added smooth argument and tested with [nltk](https://nltk.org).

Borda commented on requirements/test.txt:
```diff
@@ -12,3 +12,4 @@ black==19.10b0
 pre-commit>=1.0
 
 cloudpickle>=1.2
+nltk
```
do we have a version limitation?

ydcjeff replied:
okay, will limit with current latest version.

Borda replied:
it is fine for our CI, but for other users...
you can just send a commit with 1year / 2year old nltk version and see if tests are still passing :]

ydcjeff replied:
Sorry about that as I am always used to latest version 😁

I have locked with `nltk==3.3` as tests failed on `3.2`.

Borda replied:
I agree that it is better but some users may freeze their version because of some functional requirements... :]
so let's make it `nltk>=3.3`

ydcjeff replied:
Done it.

Borda commented on tests/metrics/functional/test_sequence.py:
```diff
@@ -0,0 +1,206 @@
+def test_with_sentence_bleu():
+    nltk_output = sentence_bleu([reference1, reference2, reference3], hypothesis1, weights=(1, 0, 0, 0))
+    pl_output = bleu_score([hypothesis1], [[reference1, reference2, reference3]], n=1).item()
+    assert round(pl_output, 4) == round(nltk_output, 4)
```
use `torch.allclose(...)`

Borda commented on tests/metrics/functional/test_sequence.py:
```diff
@@ -0,0 +1,206 @@
+smooth_func = SmoothingFunction().method2
+
+# example taken from https://www.nltk.org/api/nltk.translate.html?highlight=bleu%20score#nltk.translate.bleu_score.sentence_bleu
+hypothesis1 = "It is a guide to action which ensures that the military always obeys the commands of the party".split(' ')
```
pls use this format also for other sentences bellow...

Borda replied:
as this is kind of Constant, rather use capitals? @awaelchli

Borda commented on tests/metrics/functional/test_sequence.py:
```diff
@@ -0,0 +1,206 @@
+
+
+def test_with_sentence_bleu():
+    nltk_output = sentence_bleu([reference1, reference2, reference3], hypothesis1, weights=(1, 0, 0, 0))
```
this looks very similar, you are testing the same just with different `weights` and `n`, right?
if so use parameterization so it is easier to see that it is the same...

ydcjeff replied:
Yea, I have parameterized it.

Should I also parametrize `smooth` argument?
Right now, I am testing with 4 different `weights` and `n` with `smooth=True` and `smooth=False`.

Borda replied:
yes, I missed the `smooth` :rabbit:

Borda commented on tests/metrics/functional/test_sequence.py:
```diff
@@ -0,0 +1,206 @@
+    assert round(pl_output, 4) == round(nltk_output, 4)
+
+
+hyp = [[]]
+ref = [[[]]]
+
+
+def test_empty_bleu():
+    assert bleu_score(hyp, ref) == 0.0
+
+
+hyps = [["My", "full", "pytorch"]]
+refs = [[["My", "full", "pytorch", "test"], ["Completely", "Different"]]]
+
+
+def test_no_4_gram():
+    assert bleu_score(hyps, refs) == 0.0
```
```suggestion
def test_empty_bleu():
    hyp = [[]]
    ref = [[[]]]
    assert bleu_score(hyp, ref) == 0.0


def test_no_4_gram():
    hyps = [["My", "full", "pytorch"]]
    refs = [[["My", "full", "pytorch", "test"], ["Completely", "Different"]]]
    assert bleu_score(hyps, refs) == 0.0
```

Borda commented on pytorch_lightning/metrics/functional/sequence.py:
```diff
@@ -0,0 +1,85 @@
+
+    Example:
+
+        >>> t = [["the", "FAST", "brown", "fox", "jumped", "over", "the", "lazy", "dog"]]
+        >>> r = [[["the", "quick", "brown", "fox", "jumped", "over", "the", "lazy", "dog"], ["the", "quick", "brown", "fox", "jumped", "over", "the", "the", "lazy", "cat"]]]
```
```suggestion
        >>> t = ["the FAST brown fox jumped over the lazy dog".split(' ')]
        >>> r = [["the quick brown fox jumped over the lazy dog".split(' '), "the quick brown fox jumped over the the lazy cat".split(' ')]]
```
rather also use string sentence

ydcjeff replied:
Okay, I totally missed that.

ydcjeff commented:
Should I start working on the module interface?

Borda commented:
@ydcjeff is it still wip or ready to review/go?

ydcjeff commented:
@Borda module interface is WIP,
I think functional interface is done.

ydcjeff commented:
I think It is now ready to review.

justusschock commented on pytorch_lightning/metrics/functional/sequence.py:
```diff
@@ -0,0 +1,85 @@
+    precision_scores = torch.zeros(n_gram)
+    c = 0.0
+    r = 0.0
+    # referenced from https://pytorch.org/text/_modules/torchtext/data/metrics.html#bleu_score
```
just saw this comment. If there is already a metric implemented in torch, why don't we just import this with an optional import here? Torchtext is part of the Pytorch environment, so this should be find, right @Borda ?

ydcjeff replied:
The implementation is almost similar, the only differences are 
- current implementation includes `smooth` argument, torchtext doesn't
- current implementation returns `Tensor`, torchtext returns normal `float`

Borda replied:
Torchtext is extras so we can make option I am fine with that =)

ydcjeff replied:
So, shall I refactor with torchtext?

Borda replied:
If we refactor it with torchtext, users without it cannot use this metric, but this metric is specific for text so they most likely have torchtext anyway... So let's refactor it...
Cc: @justusschock

ydcjeff commented:
I have refactored with torchtext.

ydcjeff commented:
any progress so far?

Borda commented on pytorch_lightning/metrics/functional/sequence.py:
```diff
@@ -0,0 +1,29 @@
+from typing import List
+
+import torch
+from torchtext.data.metrics import bleu_score as bleu
```
this import shall be optional, pls check how it is done with sklearn metrics or loggers

ydcjeff replied:
@Borda torchtext had changed to optional import

Borda commented on pytorch_lightning/metrics/functional/sequence.py:
```diff
@@ -0,0 +1,37 @@
+
+try:
+    from torchtext.data.metrics import bleu_score as bleu
+
+    _TORCHTEXT_AVAILABLE = True
+except ImportError:  # pragma: no-cover
+    _TORCHTEXT_AVAILABLE = False
```
```suggestion
except ImportError:  # pragma: no-cover
    TORCHTEXT_AVAILABLE = False
else:
    TORCHTEXT_AVAILABLE = True
```

ydcjeff replied:
Done!

Borda commented on pytorch_lightning/metrics/functional/sequence.py:
```diff
@@ -0,0 +1,37 @@
+    """
+    if not _TORCHTEXT_AVAILABLE:
+        raise ImportError(
+            "Using BLEU Score Metric requires `torchtext` to be installed,"  # pragma: no-cover
```
```suggestion
            "Using BLEU Score Metric requires `torchtext` to be installed,"
```

ydcjeff replied:
Changes Done

Borda commented on tests/metrics/functional/test_sequence.py:
```diff
@@ -0,0 +1,65 @@
+
+# example taken from
+# https://www.nltk.org/api/nltk.translate.html?highlight=bleu%20score#nltk.translate.bleu_score.sentence_bleu
+hypothesis1 = "It is a guide to action which ensures that the military always obeys the commands of the party".split()
```
As they are constants, pls use CAPITALS in name

ydcjeff replied:
Changes Done

Borda commented on tests/metrics/functional/test_sequence.py:
```diff
@@ -0,0 +1,65 @@
+        pytest.param(4, [1, 2, 3, 4]),
+    ],
+)
+class TestBLEUScore:
```
no need for class

ydcjeff replied:
Then how to parametrize two functions

ydcjeff replied:
I figured out.

justusschock commented on pytorch_lightning/metrics/functional/sequence.py:
```diff
@@ -0,0 +1,37 @@
+    from torchtext.data.metrics import bleu_score as bleu
+except ImportError:  # pragma: no-cover
+    _TORCHTEXT_AVAILABLE = False
+else:
```
@Borda why should this be in `else`? Can't we do this after the import from torchtext? IMO this would be cleaner and easier to read

Borda replied:
the else separate what is the code block potential to fail and what the actions according to the results

justusschock commented on pytorch_lightning/metrics/functional/sequence.py:
```diff
@@ -0,0 +1,37 @@
+    if not _TORCHTEXT_AVAILABLE:
+        raise ImportError(
+            "Using BLEU Score Metric requires `torchtext` to be installed,"
+            " install it with `conda install -c pytorch torchtext`."
```
maybe recommend pip here? not everyone uses conda environments and pip is basically shipped with every type of python env (also virtualenv and stuff)

ydcjeff replied:
Okay, I will change to `pip`

justusschock commented on pytorch_lightning/metrics/sequence.py:
```diff
@@ -0,0 +1,46 @@
+        Return:
+            float: BLEU Score
+        """
+        return bleu_score(
```
can we cast the output to a tensor and put it to `self.device`?

ydcjeff replied:
Does it also need to cast the output to a tensor for functional metric?

justusschock replied:
the issue is that for functionals we don't know the device

ydcjeff replied:
yea, I just cast in the Class metric

justusschock commented on pytorch_lightning/metrics/sequence.py:
```diff
@@ -0,0 +1,46 @@
+        self.n_gram = n_gram
+        self.weights = weights
+
+    def forward(self, translate_corpus: List[str], reference_corpus: List[str]) -> float:
```
does this only work with List of strings or would tuples etc. work as well? If so, please change the typing to `Sequence`

Borda replied:
@ydcjeff ^^

ydcjeff commented:
Hello, anything left to change?

williamFalcon commented:
i think you have to rebase now?

ydcjeff commented:
@Borda some tests are failing as `torchtext` is in not in the `test.txt`, some are due to `UNEXPECTED EXCEPTION: ImportError...` and some ubuntu test are due to `PyYAML` install, I think

Borda commented:
`torchrext` shall be in extra.txt

Borda commented:
@ydcjeff mind check if I did not forget anything?

ydcjeff commented:
@Borda yes, all good

Borda commented:
> Borda yes, all good

cool, mind fixing the test so we can let it lend...

ydcjeff commented:
> > Borda yes, all good
> 
> cool, mind fixing the test so we can let it lend...

Yes, I am fixing now

ydcjeff commented:
@Borda 
[this circleci test](https://app.circleci.com/pipelines/github/PyTorchLightning/pytorch-lightning/6563/workflows/937b2c62-da4b-43fd-9828-0d8c1a0a6970/jobs/48422/steps) failed as there is only installing from `base.txt` and `test.txt` while `torchtext` is in `extra.txt`

Borda commented on CHANGELOG.md:
```diff
@@ -9,6 +9,7 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 
 ### Added
 
+- Added BLEU Score Class and Functional Metrics ([#2535](https://github.com/PyTorchLightning/pytorch-lightning/pull/2535))
```
```suggestion
- Added BLEU metrics ([#2535](https://github.com/PyTorchLightning/pytorch-lightning/pull/2535))
```

Borda commented on docs/source/metrics.rst:
```diff
@@ -28,7 +28,7 @@ Example::
 .. warning::
     The metrics package is still in development! If we're missing a metric or you find a mistake, please send a PR!
-    to a few metrics. Please feel free to create an issue/PR if you have a proposed  metric or have found a bug.
+    to a few metrics. Please feel free to create an issue/PR if you have a proposed metric or have found a bug.
```
it seems to be something missing here...

ydcjeff replied:
> The metrics package is still in development! If we’re missing a metric or you find a mistake, please send a PR! to a few metrics. Please feel free to create an issue/PR if you have a proposed metric or have found a bug.
Same in the latest docs.

Borda replied:
it has been added in #2209  @SkafteNicki mind check what you had in mind? :]

Borda commented on pytorch_lightning/metrics/sequence.py:
```diff
@@ -0,0 +1,51 @@
+
+    Example:
+
+        >>> translate_corpus = ['the cat is on the mat'.split()]
```
@awaelchli can we also add here a conditional skip in case that `torchtext` is not installed?

awaelchli replied:
No, for pytest doctesting this is not possible. 
Conditional skipping can only be done with sphinx doctests. 
I suggest to install torchtext in the doctest CI job or rewrite the test without torchtext

Borda replied:
well the purpose of these test it to check if the package can run without extras, so install extras extra goes against this test :]
so just add this module to ignore for the particular test... (as sklearns already is)

ydcjeff replied:
Should I add `--ignore=pytorch_lightning/metrics/sequence.py`?

Borda commented:
> [this circleci test](https://app.circleci.com/pipelines/github/PyTorchLightning/pytorch-lightning/6563/workflows/937b2c62-da4b-43fd-9828-0d8c1a0a6970/jobs/48422/steps) failed as there is only installing from `base.txt` and `test.txt` while `torchtext` is in `extra.txt`

I thought that we are renaming this sequence module to `nlp`? @justusschock 
you shall add this module do ignore for Conda and CIrcleCI tests :]

ydcjeff commented:
> I thought that we are renaming this sequence module to `nlp`? @justusschock

so `nlp` or `sequence`?

ydcjeff commented:
> you shall add this module do ignore for Conda and CIrcleCI tests :]

how shall I add?

williamFalcon commented:
this should be:

from p...metrics.nlp import Bleu

ydcjeff commented:
> this should be:
> 
> from p...metrics.nlp import Bleu

Okay

ydcjeff commented:
I guess this is ready to review. @williamFalcon @Borda

williamFalcon commented:
ummm. so this is a wrapper on torchtext? i think we need our own implementation

ydcjeff commented:
Ahhh, I implemented from scratch and referenced the implmentation a little from torchtext.

Then, @justusschock sugguested to base on torchtext since it is in pytorch ecosystem.

So I refactored with torchtext.
@williamFalcon

justusschock commented:
@williamFalcon : @Borda and me agreed, that we don't need to duplicate this if it is already present within torchtext since basically everyone that will use this will also have torchtext installed and it was already an optional dependency.

williamFalcon commented:
i guess the point of metrics here is to centralize all metrics. otherwise we coule have said the same about sklearn. 

we want our metrics package to be the reference implementation for any metric

justusschock commented:
With sklearn it's more about GPU performance/syncs :)

But I see your point. Then I'm sorry @ydcjeff :D But your code should still be available here :) SO just copy/paste it in :D

ydcjeff commented:
@williamFalcon I have re-implemented from scratch, anything you like to add on?

awaelchli commented on pytorch_lightning/metrics/functional/nlp.py:
```diff
@@ -0,0 +1,88 @@
+    precision_scores = torch.zeros(n_gram)
+    c = 0.0
+    r = 0.0
+    # referenced from https://pytorch.org/text/_modules/torchtext/data/metrics.html#bleu_score
```
If this code was adapted from torchtext, the  reference should probably go to the docstring or top of the file and with proper attribution with name of library, authors, date and link

awaelchli commented on .circleci/config.yml:
```diff
@@ -19,7 +19,7 @@ references:
      command: |
        python --version ; pip --version ; pip list
-       python -m pytest pytorch_lightning -v --junitxml=test-reports/pytest_junit.xml  --ignore=pytorch_lightning/loggers/comet.py --ignore=pytorch_lightning/loggers/mlflow.py --ignore=pytorch_lightning/loggers/neptune.py --ignore=pytorch_lightning/loggers/test_tube.py --ignore=pytorch_lightning/loggers/wandb.py --ignore=pytorch_lightning/metrics/sklearns.py
+       python -m pytest pytorch_lightning -v --junitxml=test-reports/pytest_junit.xml  --ignore=pytorch_lightning/loggers/comet.py --ignore=pytorch_lightning/loggers/mlflow.py --ignore=pytorch_lightning/loggers/neptune.py --ignore=pytorch_lightning/loggers/test_tube.py --ignore=pytorch_lightning/loggers/wandb.py --ignore=pytorch_lightning/metrics/sklearns.py --ignore=pytorch_lightning/metrics/nlp ...
```
is the ignore still needed now that torchtext is not used in docstring?

ydcjeff replied:
Ohh yea, I forgot to remove them

ydcjeff commented:
It's ready to be reviewed

justusschock commented on pytorch_lightning/metrics/functional/nlp.py:
```diff
@@ -0,0 +1,92 @@
+import torch
+
+
+def _count_ngram(ngram_input_list: list, n_gram: int) -> Counter:
```
```suggestion
def _count_ngram(ngram_input_list: List[str], n_gram: int) -> Counter:
```

justusschock commented on pytorch_lightning/metrics/functional/nlp.py:
```diff
@@ -0,0 +1,92 @@
+# Authors: torchtext authors and @sluks
+# Date: 2020-07-18
+# Link: https://pytorch.org/text/_modules/torchtext/data/metrics.html#bleu_score
+from typing import Sequence
```
```suggestion
from typing import Sequence, List
```

justusschock commented on pytorch_lightning/metrics/functional/nlp.py:
```diff
@@ -0,0 +1,92 @@
+
+
+def bleu_score(
+    translate_corpus: Sequence, reference_corpus: Sequence, n_gram: int = 4, smooth: bool = False
```
```suggestion
    translate_corpus: Sequence[str], reference_corpus: Sequence[str], n_gram: int = 4, smooth: bool = False
```

justusschock commented on pytorch_lightning/metrics/nlp.py:
```diff
@@ -0,0 +1,46 @@
+            reference_corpus=reference_corpus,
+            n_gram=self.n_gram,
+            smooth=self.smooth,
+        )
```
```suggestion
        ).to(self.device, self.dtype)
```

williamFalcon commented:
@ydcjeff awesome!!
